src/database-postgresql.cpp

```diff
@@ -0,0 +1,319 @@
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "config.h"
+
+#if USE_POSTGRESQL
+
+#include "database-postgresql.h"
+#include "log.h"
+#include "filesys.h"
+
+
+Database_PostgreSQL::Database_PostgreSQL(Settings &conf)
+{
+	m_conf = conf;
+
+	if (PQisthreadsafe() < 1)
+		infostream << "Postgres backend not thread safe." << std::endl;
+
+	connectString();
+	dbconnect();
+	prepare();
+}
+
+
+void Database_PostgreSQL::beginSave()
+{
+	verifyDatabase();
+	resultsCheck(PQexec(m_conn, "BEGIN;"));
+}
+
+
+void Database_PostgreSQL::endSave()
+{
+	resultsCheck(PQexec(m_conn, "COMMIT;"));
+}
+
+
+bool Database_PostgreSQL::deleteBlock(const v3s16 &pos)
+{
+	return execPrepared("delete", 3, pos);
+}
+
+
+bool Database_PostgreSQL::saveBlock(const v3s16 &pos, const std::string &data)
+{
+	if (m_pgVersion < 90500) {
+		execPrepared("update", 4, pos, false, data);
+
+		if (PQcmdTuples(m_res)) {
+			PQclear(m_res);
+			return true;
+		}
+	}
+
+	return execPrepared("save", 4, pos, true, data);
+}
+
+
+std::string Database_PostgreSQL::loadBlock(const v3s16 &pos)
+{
+	std::string data;
+
+	execPrepared("load", 3, pos, false, data, PGRES_TUPLES_OK);
+
+	if (PQntuples(m_res)) {
+		data = std::string(PQgetvalue(m_res, 0,0),
+			PQgetlength(m_res, 0, 0));
+	}
+
+	PQclear(m_res);
+	return data;
+}
+
+
+void Database_PostgreSQL::listAllLoadableBlocks(std::vector<v3s16> &dst)
+{
+	resultsCheck(PQexec(m_conn, "SELECT x, y, z FROM blocks;"), false,
+		PGRES_TUPLES_OK);
+
+	int numrows = PQntuples(m_res);
+
+	if (!numrows) {
+		PQclear(m_res);
+		return;
+	}
+
+	dst.reserve(numrows);
+
+	for (int row = 0; row < numrows; row++) {
+		dst.push_back(v3s16(
+			atoi(PQgetvalue(m_res, row, 0)),
+			atoi(PQgetvalue(m_res, row, 1)),
+			atoi(PQgetvalue(m_res, row, 2))));
+	}
+
+	PQclear(m_res);
+}
+
+
+void Database_PostgreSQL::connectString()
+{
+	if (m_conf.exists("pg_connection_info")) {
+		m_conninfo = m_conf.get("pg_connection_info");
+	} else {
+		infostream << "PostgreSQL: pg_connection_info not found in";
+		infostream << "world.mt file. Using defaul datbase name ";
+		infostream << "minetest." << std::endl;
+		m_conninfo = std::string("dbname=minetest");
+	}
+
+	m_conninfo.append(" application_name=minetest");
+
+	infostream << "PostgreSQL: connection info: " << m_conninfo;
+	infostream << std::endl;
+}
+
+
+void Database_PostgreSQL::dbconnect()
+{
+	m_conn = PQconnectdb(m_conninfo.c_str());
+
+	if (PQstatus(m_conn) != CONNECTION_OK) {
+		throw FileNotGoodException(std::string(
+			"PostgreSQL database error: ") +
+			PQerrorMessage(m_conn));
+	}
+
+	m_pgVersion = PQserverVersion(m_conn);
+
+	if (m_pgVersion < 90300) {
+		throw FileNotGoodException("PostgreSQL database error: "
+			"Server version 9.3 or greater required.");
+	}
+
+	createDatabase();
+
+	infostream << "PostgreSQL datbase: Server version " << m_pgVersion;
+	infostream << std::endl;
+	infostream << "PostgreSQL: Database connection made." << std::endl;
```

Comment:
Should be `verbosestream`, and you only have to mention the stream once:
```C++
verbosestream << "Version " << version << std::endl
	<< "Connection made" << std::endl;
```
Also: `datbase`->`database` again!